qiskit/primitives/containers/data_bin.py

```diff
@@ -54,10 +55,30 @@ def __repr__(self):
         vals = (f"{name}={getattr(self, name)}" for name in self._FIELDS if hasattr(self, name))
         return f"{type(self)}({', '.join(vals)})"
 
+    def __getitem__(self, key: str) -> Any:
+        if not hasattr(self, key):
+            raise ValueError(f"Key ({key}) does not exist in this data bin.")
+        return getattr(self, key)
+
+    def __contains__(self, key: str) -> bool:
+        return hasattr(self, key)
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self._FIELDS)
+
+    def keys(self) -> Iterable[str]:
+        """Return an iterable of field names."""
+        return iter(self)
```

Comment:
I tried `keys` and confirmed that the return type `dict_keys` supports `len` but does not support `getitem`. But, since data bin is immutable, I think a list is a good choice.

```python
In [1]: a = {1: 2, 2: 3}
In [2]: keys = a.keys()

In [3]: type(keys)
Out[3]: dict_keys

In [4]: len(keys)
Out[4]: 2

In [5]: keys[0]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 keys[0]

TypeError: 'dict_keys' object is not subscriptable
```